[Messenger] Allow to close the transport connection

andrew-demb opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #53543 
| License       | MIT

<!--
Replace this notice by a description of your feature/bugfix.
This will help reviewers and should be a good start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against the latest branch.
 - For new features, provide some code snippets to help understand usage.
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
 - Never break backward compatibility (see https://symfony.com/bc).
-->

~~1. Implemented the possibility to make messenger transports resettable~~
~~2. Implemented reset Redis connection for Redis messenger transport~~

~~This feature may lead to decreased performance for messenger consumers (because connection will be resetted between processing messages).~~

~~One way to resolve it that I found - add configuration for resettable transports - aka "reset connection on kernel reset: true/false". For consumers it can be configured via env var to be "false", but for web "true".~~

----

**UPD 2025-02-27**: According to the feedback, I changed the implementation to another one to help with our use case.

Implemented a way to close messenger transport from the application, to allow free resources for long-running processes (like a long-running webserver)

andrew-demb committed:
[Messenger] Allow to close the transport connection

andrew-demb commented:
> Error: src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:59:13: UndefinedClass: Class, interface or enum named Relay\Relay does not exist (see https://psalm.dev/019)
> Error: src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:740:34: UndefinedClass: Class, interface or enum named Relay\Relay does not exist (see https://psalm.dev/019)

I don't know how to handle this error. I didn't add Relay in the PR - probably an issue introduced by the Psalm update or existed before.

andrew-demb commented:
> .................................zend_mm_heap corrupted
> Job exited with: 134
> Error: KO src/Symfony/Component/Validator

This CI failure for PHP 8.5 seems unrelated to PR

ro0NL commented:
doctrine can be optionally closed at the middelware layer: https://github.com/symfony/symfony/blob/7.3/src/Symfony/Bridge/Doctrine/Messenger/DoctrineCloseConnectionMiddleware.php

im wondering if redis should follow a similar approach

andrew-demb commented:
I can introduce the interface "CloseableTransportInterface" that will allow closing used transport connections from the userland (in my case - from "kernel.terminate" event listener).

I can see, that this can be implemented by most transports (AMQP, Redis, Doctrine DBAL, etc)

andrew-demb commented:
According to the feedback, I changed the implementation to another one to help with our use case.

Implemented a way to close messenger transport from the application, to allow free resources for long-running processes (like a long-running webserver)

stof commented on src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php:
```diff
@@ -80,6 +80,12 @@ public function reset(): void
 
+    public function close(): void
+    {
+        $this->driverConnection->close();
```
this transport does not own the DBAL connection, so it should not close it either

andrew-demb replied:
Updated. Other transports seem OK to me to be closeable (they remain their own connection or already have some "reset" methods).

ro0NL replied:
@stof from a messenger:consume POV i think doctrine transport is "closeable"

that's what DoctrineCloseConnectionMiddleware does no?

i tend to believe we should either solve it at the transport layer or the middleware layer in an agnostic matter IMHO

stof replied:
DoctrineCloseConnectionMiddleware is closing the connection *if you enable that middleware*.

the transport itself should not close the connection implicitly, because it is not owning the connection (the connection might be used for multiple purposes)

fabpot commented on src/Symfony/Component/Messenger/Transport/CloseableTransportInterface.php:
```diff
@@ -0,0 +1,22 @@
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
```
This should be removed as we're not using it in the Symfony code base.

andrew-demb replied:
done

fabpot commented on src/Symfony/Component/Messenger/Transport/CloseableTransportInterface.php:
```diff
@@ -0,0 +1,22 @@
+interface CloseableTransportInterface
+{
+    /**
+     * Close the transport.
```
I think it can be removed;

andrew-demb replied:
done

fabpot reviewed (changes requested):
Can you add a note in the CHANGELOG file about the new interface?

andrew-demb commented:
status:needs review

fabpot commented:
Thank you @andrew-demb.
